fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/jobs/updateloanarrearsageing/LoanArrearsAgeingUpdateHandler.java

```diff
@@ -0,0 +1,202 @@
+                + "coalesce(mr.penalty_charges_waived_derived, 0) - coalesce(mr.penalty_charges_completed_derived, 0))";
+
+        updateSqlBuilder.append(
+                "INSERT INTO m_loan_arrears_aging(loan_id,principal_overdue_derived,interest_overdue_derived,fee_charges_overdue_derived,penalty_charges_overdue_derived,total_overdue_derived,overdue_since_date_derived)");
+        updateSqlBuilder.append("select ml.id as loanId,");
+        updateSqlBuilder.append(principalOverdueCalculationSql + " as principal_overdue_derived,");
+        updateSqlBuilder.append(interestOverdueCalculationSql + " as interest_overdue_derived,");
+        updateSqlBuilder.append(feeChargesOverdueCalculationSql + " as fee_charges_overdue_derived,");
+        updateSqlBuilder.append(penaltyChargesOverdueCalculationSql + " as penalty_charges_overdue_derived,");
+        updateSqlBuilder.append(principalOverdueCalculationSql + "+" + interestOverdueCalculationSql + "+");
+        updateSqlBuilder.append(feeChargesOverdueCalculationSql + "+" + penaltyChargesOverdueCalculationSql + " as total_overdue_derived,");
+        updateSqlBuilder.append("MIN(mr.duedate) as overdue_since_date_derived ");
+        updateSqlBuilder.append(" FROM m_loan ml ");
+        updateSqlBuilder.append(" INNER JOIN m_loan_repayment_schedule mr on mr.loan_id = ml.id ");
+        updateSqlBuilder.append(" left join m_product_loan_recalculation_details prd on prd.product_id = ml.product_id ");
+        updateSqlBuilder.append(" WHERE ml.loan_status_id = 300 ");// active
+        if (!isForAllLoans) {
+            updateSqlBuilder.append(" and ml.id IN (?)");
+        }
+        updateSqlBuilder.append(" and mr.completed_derived is false ");
+        updateSqlBuilder.append(" and mr.duedate < ")
+                .append(sqlGenerator.subDate(sqlGenerator.currentBusinessDate(), "COALESCE(ml.grace_on_arrears_ageing, 0)", "day"))
+                .append(" ");
+        updateSqlBuilder.append(" and (prd.arrears_based_on_original_schedule = false or prd.arrears_based_on_original_schedule is null) ");
+        updateSqlBuilder.append(" GROUP BY ml.id");
+        return updateSqlBuilder.toString();
+    }
+
+    private List<String> updateLoanArrearsAgeingDetailsWithOriginalSchedule(List<Long> loanIdsForUpdate) {
+        List<String> insertStatement = new ArrayList<>();
+        String sqlForLoanIdentifiers = buildQueryForLoanIdentifiersWithOriginalSchedule(Boolean.FALSE);
+        List<Object> loanIdsForQuery = new ArrayList<>();
+        for (Long loanId : loanIdsForUpdate) {
+            loanIdsForQuery.add(loanId);
+        }
+        List<Long> loanIds = this.jdbcTemplate.queryForList(sqlForLoanIdentifiers, loanIdsForQuery.toArray(), new int[] { Types.BIGINT },
+                Long.class);
+        if (!loanIds.isEmpty()) {
+            String loanIdsAsString = loanIds.toString();
```

Comment:
This definitely is not going to fly with me. Calling the toString on the List and then cutting it off to be appended to the SQL string, that's purely wrong. Use parameterized queries with proper types.

Reply:
Was existing code which was not modified. Refactored now.